vochain/indexer/indexer.go

```diff
@@ -249,37 +254,112 @@ func (idx *Indexer) RestoreBackup(path string) error {
+	}
+	log.Debugf("restoring indexer backup")
+	gzipReader, err := gzip.NewReader(r)
+	if err != nil {
+		return fmt.Errorf("could not create gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+	if err := restoreDBFromSQLDump(idx.dbPath, gzipReader); err != nil {
+		return fmt.Errorf("could not restore indexer backup: %w", err)
+	}
+	if err := idx.startDB(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func restoreDBFromSQLDump(dbPath string, r io.Reader) error {
+	db, err := sql.Open("sqlite3", fmt.Sprintf("file:%s?mode=rwc&_journal_mode=wal&_txlock=immediate&_synchronous=normal&_foreign_keys=true", dbPath))
+	if err != nil {
+		return fmt.Errorf("could not open indexer db: %w", err)
+	}
+	defer db.Close()
+
+	scanner := bufio.NewScanner(r)
+	var statement strings.Builder
+	for scanner.Scan() {
+		line := scanner.Text()
+		statement.WriteString(line)
+		statement.WriteString("\n")
+
+		if strings.HasSuffix(line, ";") {
+			_, err := db.Exec(statement.String())
+			if err != nil {
+				return fmt.Errorf("failed to execute statement: %s (error: %w)", statement.String(), err)
+			}
+			statement.Reset()
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("error during restore: %w", err)
+	}
+
+	return nil
+}
+
+// ExportBackup backs up the database to a file on disk.
 // Note that writes to the database may be blocked until the backup finishes,
 // and an error may occur if a file at path already exists.
 //
 // For sqlite, this is done via "VACUUM INTO", so the resulting file is also a database.
-func (idx *Indexer) SaveBackup(ctx context.Context, path string) error {
-	_, err := idx.readOnlyDB.ExecContext(ctx, `VACUUM INTO ?`, path)
-	return err
+func (idx *Indexer) ExportBackup(w io.Writer) error {
+	log.Debugf("exporting indexer backup")
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	tmpDB, err := os.CreateTemp("", "indexer*.sqlite3")
+	if err != nil {
+		return fmt.Errorf("could not create tmpdb file: %w", err)
+	}
+	defer func() {
+		if err := os.Remove(tmpDB.Name()); err != nil {
+			log.Warnw("error removing tmpdb file", "path", tmpDB.Name(), "err", err)
+		}
+	}()
+
+	if _, err := idx.readOnlyDB.ExecContext(ctx, `VACUUM INTO ?`, tmpDB.Name()); err != nil {
+		return fmt.Errorf("could not vacuum into tmpdb: %w", err)
+	}
+
+	db, err := sql.Open("sqlite3", tmpDB.Name())
+	if err != nil {
+		return fmt.Errorf("could not open tmpDB: %w", err)
+	}
+	defer db.Close()
+
+	// first drop stats table
+	if _, err := db.ExecContext(ctx, `DROP TABLE IF EXISTS sqlite_stat1;`); err != nil {
+		return fmt.Errorf("could not drop table sqlite_stat1: %w", err)
+	}
+
+	// make goose_db_version table deterministic
+	if _, err := db.ExecContext(ctx, `UPDATE goose_db_version SET tstamp = '1970-01-01 00:00:00';`); err != nil {
+		return fmt.Errorf("could not update goose_db_version: %w", err)
+	}
+
+	gzw := gzip.NewWriter(w)
```

Comment:
we already use zstd, so why not use it here as well? it's better than gzip in every way :)

Reply:
yes, there is a compressor package already